### ISSUE TITLE
HttpClient() response leak

### ISSUE THREAD

**Issue by icanhasjonas** [elastic/elasticsearch-net#2784]
**NEST/Elasticsearch.Net version**:

**Elasticsearch version**: 5.4

**Description of the problem including expected versus actual behavior**:
It would seem that the response from the HttpClient is never dispose and therefore producing a leak. The linux implementation is using cURL and that seems to produce the notorious `Too many open files` exception.

CODE: https://github.com/elastic/elasticsearch-net/blob/master/src/Elasticsearch.Net/Connection/HttpConnection-CoreFx.cs#L70-L78

**Steps to reproduce**:
 1. Run for a long time, on linux, where the HttpClient implementation is Curl based
 2.
 3.

**Provide `ConnectionSettings` (if relevant)**:

**Provide `DebugInformation` (if relevant)**:

**Describe the feature**:

**Comment by Mpdreamz** [comment]
The underlying stream is disposed but later on as part of the interplay between ResponseBuilder and Serializer. 

https://github.com/elastic/elasticsearch-net/blob/master/src/Tests/ClientConcepts/ConnectionPooling/Dispose/ResponseBuilderDisposeTests.cs

Which is what response dispose ultimately does as as well:

https://github.com/dotnet/corefx/blob/master/src/System.Net.Http/src/System/Net/Http/HttpResponseMessage.cs#L203-L215
https://github.com/dotnet/corefx/blob/master/src/System.Net.Http/src/System/Net/Http/HttpContent.cs#L468

We also have integration tests that tests we reuse connections properly. 

That said integration tests do not run continuously on our Linux CI (only unit tests), and our CI runs the integration tests on desktop CLR so will dive in tomorrow to see if I can reproduce the connection leak there.

Thanks for reporting @icanhasjonas

**Comment by icanhasjonas** [comment]
Here's the file you should be looking at: https://github.com/dotnet/corefx/blob/release/1.1.0/src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.CurlResponseMessage.cs

.. the `HttpResponseMessage` needs to be disposed for the `CurlResponseStream` to dispose.. 

https://github.com/dotnet/corefx/blob/release/1.1.0/src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.CurlResponseMessage.cs#L24-L25

I wrote a small patch for the offending code, and deployed in production and our `Too many open files` issues havn't been seen since.

**Comment by icanhasjonas** [comment]
Here's a workaround HttpClientConnection() for now.. 
https://gist.github.com/icanhasjonas/3002659400b1c706da87ccd31bb46806

**Comment by Mpdreamz** [comment]
Thanks for posting that up Jonas, I've been working on a fix and getting
the integration tests to run on Linux. This brought up another problem that
our integration tests for connection reuse indicate CurlHandler is not
reusing tcp connections despite sending KeepAlive. This might explain some
of the issues we got saying it's slower.

Now on a goose chase trying to compile CurlHandler and setting curl tcp
keep alive flags.

**Comment by icanhasjonas** [comment]
Thanks,

Here's some more information on the lifetime management of the objects in the cURL world, https://github.com/dotnet/corefx/blob/4709a5cda50ec3d141b5b460bc05c8b576332659/src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.cs#L84-L87

**Comment by Mpdreamz** [comment]
Read that yeah :+1: 

I've found a fix for the 2nd problem where Elasticsearch Stats API will report too many total http connection have been opened.
 [We have a test for this that asserts it never goes above 100](https://github.com/elastic/elasticsearch-net/blob/master/src/Tests/Reproduce/ConnectionReuseAndBalancing.cs#L70) total http connections opened on a fresh cluster even after firing `10.000` request at it. 10k is just to keep the total running time of the test to a minimum, increasing would still keep that number at a value < `100` total opened connections. The clients default concurrent connection limit is `80` and the actual value of that tests will typically be `80-85`, taking into account connections from cluster setup which may not be reused depending on timing.

This test consistently fails on .net core on linux which means it opens and closes a lot of sockets quickly exceeding a factor 2 of the `ConnectionLimit` setting in the client. 

The only way to keep it stable is to set the `ConnectionLimit` to the number of physical processors in a machine (`lscpu -p | egrep -v '^#' | sort -u -t, -k 2,4 | wc -l`).  

On my machine with `4` physical processors using that is the only way to keep `TotalOpened` and `CurrentOpened` at a stable number no matter if I do `10k` or `100k` searches.

If I print these numbers after each iteration of a 1000 searches for 

### `ConnectionLimit(4)`

```
Current Open: 4, Total Opened: 5
(repeated 20 times)
```

**total runtime:** 44seconds


### `ConnectionLimit(8)` (20 iterations)

```
Current Open: 4, Total Opened: 9
Current Open: 7, Total Opened: 13
Current Open: 8, Total Opened: 15
Current Open: 7, Total Opened: 16
Current Open: 7, Total Opened: 18
Current Open: 8, Total Opened: 20
Current Open: 7, Total Opened: 21
Current Open: 8, Total Opened: 23
Current Open: 8, Total Opened: 24
Current Open: 7, Total Opened: 25
Current Open: 7, Total Opened: 27
Current Open: 8, Total Opened: 29
Current Open: 8, Total Opened: 30
Current Open: 8, Total Opened: 31
Current Open: 7, Total Opened: 32
Current Open: 7, Total Opened: 34
Current Open: 7, Total Opened: 36
Current Open: 7, Total Opened: 38
Current Open: 8, Total Opened: 40
Current Open: 7, Total Opened: 41
Current Open: 8, Total Opened: 43
```

**total runtime:** 31seconds


### The default `ConnectionLimit(80)` (20 iterations)

```
Current Open: 4, Total Opened: 11
Current Open: 73, Total Opened: 87
Current Open: 80, Total Opened: 95
Current Open: 80, Total Opened: 102
Current Open: 80, Total Opened: 115
Current Open: 72, Total Opened: 116
Current Open: 79, Total Opened: 125
Current Open: 80, Total Opened: 127
Current Open: 79, Total Opened: 128
Current Open: 68, Total Opened: 132
Current Open: 68, Total Opened: 145
Current Open: 80, Total Opened: 184
Current Open: 80, Total Opened: 186
Current Open: 79, Total Opened: 206
Current Open: 80, Total Opened: 222
Current Open: 78, Total Opened: 278
Current Open: 70, Total Opened: 284
Current Open: 66, Total Opened: 302
Current Open: 80, Total Opened: 317
Current Open: 64, Total Opened: 322
Current Open: 79, Total Opened: 342
```

**total runtime:** 25seconds

-----

I *suspect* this property is at play in combination with thread preempting:

https://github.com/dotnet/corefx/blob/bffef76f6af208e2042a2f27bc081ee908bb390b/src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.MultiAgent.cs#L32


I think for now a safer default for `ConnectionLimit` on .NET core linux is `Environment.ProcessorCount` in terms of performance vs socket bleed. 

I need to raise this on https://github.com/dotnet/corefx/ as well since the characteristics between platforms seems way off being able to sustain 80 vs 4 concurrent requests without socket bleed.

**Comment by Mpdreamz** [comment]
Should state that (for anyone reading this in the future) The actual open file handlers will be something greater than `Current Open` taking into account those sockets in `TIME_WAIT` or `CLOSE_WAIT` state.

**Comment by Mpdreamz** [comment]
This should now be fixed as #2785 

Opened https://github.com/dotnet/corefx/issues/21227 to track greater TCP reuse issue.
